### PR TITLE
Introduce execution context and expose operation.

### LINKED ISSUE #1770: Scenario: Add the ability to branch based on the publisher
It should be possible to determine if a manifest is being published or if the code is going to execute regardless of the environment. This would let developers change the app model based on what "mode" the app mode is being executed in.

### PR DESCRIPTION
**Pull request by mitchdenny**
Fixes #1770 and #1823 

Introduce a new execution context which can be used on the builder and via DI.

## Old approach

```csharp
.WithEnvironment(context =>
{
    if (context.PublisherName == "manifest")
    {
        context.EnvironmentVariables.Add("MSSQL_SA_PASSWORD", $"{{{sqlServer.Name}.inputs.password}}");
        return
    }
    ...
```

## New approach

```csharp
.WithEnvironment(context =>
{
    if (context.ExecutionContext.Operation == DistributedAppliationOperation.Publish)
    {
        context.EnvironmentVariables.Add("MSSQL_SA_PASSWORD", $"{{{sqlServer.Name}.inputs.password}}");
        return
    }
    ...
```

Note that the execution context is also available on the builder - example usage:

```csharp
var builder = DistributedApplication.CreateBuilder(args);
var db = builder.ExecutionContext.Operation switch {
  Publish => builder.AddConnectionString("db"),
  _ => builder.AddPostgres("pgsql").AddDatabase("db")
};
```

### CODE REVIEW

File: src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs [hunk 1]
```diff
@@ -6,17 +6,17 @@ namespace Aspire.Hosting.ApplicationModel;
...
     /// <summary>
     /// Gets the environment variables associated with the callback context.
     /// </summary>
     public Dictionary<string, string> EnvironmentVariables { get; } = environmentVariables ?? new();
 
     /// <summary>
-    /// Gets the name of the publisher associated with the callback context.
+    /// Gets the execution context associated with this invocation of the AppHost.
     /// </summary>
-    public string PublisherName { get; } = publisherName;
```
Comment (davidfowl) [inline src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs:21]: Keep PublisherName and do  this:

[Obsolete("Use ExecutionContext instead")]
string PublisherName => ExecutionContext  == Publish ? "manifest" : null;

File: src/Aspire.Hosting/DistributedApplicationRunner.cs [hunk 2]
```diff
@@ -5,23 +5,19 @@
...
     {
-        var publisherName = _options.Value.Publisher?.ToLowerInvariant() ?? "dcp";
-
-        var publisher = _serviceProvider.GetKeyedService<IDistributedApplicationPublisher>(publisherName)
-            ?? throw new DistributedApplicationException($"Could not find registered publisher '{publisherName}'");
+        var publisher = executionContext.Operation switch
+        {
+            DistributedApplicationOperation.Publish => serviceProvider.GetRequiredKeyedService<IDistributedApplicationPublisher>("manifest"),
+            _ => serviceProvider.GetRequiredKeyedService<IDistributedApplicationPublisher>("dcp")
+        };
```
Comment (davidfowl) [inline src/Aspire.Hosting/DistributedApplicationRunner.cs:19]: Undo this change. Leave the door open for publisher.

File: src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs:21]: Done.

File: src/Aspire.Hosting/DistributedApplicationRunner.cs (hunk 2 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/DistributedApplicationRunner.cs:19]: As discussed not doing this one, but leaving `PublisherName` on the envvar callback context for 1 preview.

File: src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs [hunk 3]
```diff
@@ -10,6 +10,12 @@ namespace Aspire.Hosting.ApplicationModel;
 /// <param name="environmentVariables">The environment variables associated with this execution.</param>
 public class EnvironmentCallbackContext(DistributedApplicationExecutionContext executionContext, Dictionary<string, string>? environmentVariables = null)
 {
+    /// <summary>
+    /// Obsolete. Use ExecutionContext instead. Will be removed in next preview.
+    /// </summary>
+    [Obsolete("Use ExecutionContext instead")]
+    public string PublisherName => ExecutionContext.Operation == DistributedApplicationOperation.Publish ? "manifest" : "dcp";
```
Comment (davidfowl) [inline src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs:17]:
```suggestion
    public string PublisherName => ExecutionContext.Operation == DistributedApplicationOperation.Publish ? "manifest" : null;
```

I would make it null, not dcp (was it dcp before?)

Comment (mitchdenny) [reply src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs:17]: It is "dcp".

Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/EnvironmentCallbackContext.cs:17]: Cool

File: src/Aspire.Hosting/DistributedApplicationExecutionContext.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+/// <summary>
+/// Exposes the global contextual information for this invocation of the AppHost.
+/// </summary>
+public class DistributedApplicationExecutionContext
+{
+    internal DistributedApplicationExecutionContext(DistributedApplicationOperation operation)
+    {
+        Operation = operation;
+    }
```
Comment (davidfowl) [inline src/Aspire.Hosting/DistributedApplicationExecutionContext.cs:14]: Why is this internal? How are you supposed to test these callbacks?

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 5]
```diff
@@ -8,24 +8,30 @@
...
+/// <param name="executionContext">Global contextual information for this invocation of the AppHost.</param>
+/// <param name="manifestPath">Manifest path passed in for this invocation of the AppHost.</param>
+/// <param name="writer">JSON writer used to writing the manifest.</param>
+public sealed class ManifestPublishingContext(DistributedApplicationExecutionContext executionContext, string manifestPath, Utf8JsonWriter writer)
 {
     /// <summary>
-    /// TODO: Doc Comments
+    /// Gets execution context for this invocation of the AppHost.
+    /// </summary>
+    public DistributedApplicationExecutionContext ExecutionContext { get; } = executionContext;
```
Comment (davidfowl) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:21]: Do we have any uses of this?

File: src/Aspire.Hosting/DistributedApplicationExecutionContext.cs (hunk 4 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/DistributedApplicationExecutionContext.cs]: Good point. I didn't notice because we use InternalsVisibleTo for Aspire.Hosting.Tests. Fixed.

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs (hunk 5 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:21]: Yes. As part of manifest publishing we invoke the envvar callback, which needs the executioncontext so those callbacks can do the right thing.

**Review by davidfowl — APPROVED** [review]
